### ISSUE TITLE
Make CMake API helper functions for creating VCV Rack plugin, modules, and more

### ISSUE THREAD

**Issue by nathanjhood** [StoneyDSP/StoneyVCV#121]
# Idea.

## Problem:

Currently, my `CMakeLists.txt` is only building three modules - and the plugin - but is already looking very repetitive, and will become a nuisance to maintain as the plugin complexity and number of modules grows, if done this way.

## Solution:

Create CMake helper functions which "take care" of lots of generic CMake config for `plugin` and Module targets:

All library linkages, compiler/linker configs, and so forth, would be wrapped inside the functions:

```cmake
# rack-helper-targets-cmake *or* VCVRackHelpers.cmake *or* VCVRack.cmake

function(vcvrack_add_plugin __slug __sources __version)
    cmake_parse_arguments(...)
    # This lib name is hard-coded; CMake only permits one of these per-project! ;)
    add_library(plugin SHARED)
    add_library(${__slug}::plugin ALIAS plugin)
    # validation: plugin 'VERSION_MAJOR' >= 2...
    set_target_properties(plugin PROPERTIES VERSION ${__version})
    target_sources(plugin PRIVATE ${__sources})
    target_compile_options(plugin PRIVATE "-fPIC")
    target_link_options(plugin PRIVATE "-static-libstdc++")
    target_link_libraries(${__slug}::plugin PUBLIC 
        rack::dep 
        rack::sdk 
        rack::lib
    )
    # ...and so forth!
endfunction()

function(vcvrack_add_module __name __slug __sources __version)
    cmake_parse_arguments(...)
    # This one accepts a name, but is joined to 'plugin' by the alias namespace (the 'SLUG' param)
    add_library(${__name} OBJECT)
    add_library(${__slug}::${__name} ALIAS ${__name})
    set_target_properties(plugin PROPERTIES VERSION ${__version})
    target_sources(${__name} PRIVATE ${__sources})
    target_compile_options(${__name} PRIVATE "-fPIC")
    # Since we can be sure that *all* modules must link with a target which has the name 'plugin' 
    # (since we hard-coded the name), the user just passes in the desired namespace/'SLUG' to link with :)
    target_link_libraries(${__slug}::plugin PUBLIC ${__slug}::${__name}) 
    # ...and so forth!
endfunction()
```

Then, distribute those helper functions when the VCVRack CMake targets are installed (for example, via vcpkg).

A consumer project's ideal `CMakeLists.txt` might then look like this:

```cmake
cmake_minimum_required(...)

project(StoneyVCV VERSION 2.1.107)

# The 'helpers' component would be the functions...
find_package(Rack 
    VERSION 2.5.2 
    COMPONENTS dep sdk lib helpers
)

# This would create a redistributable shared library called 'plugin.{so,dylib,dll.a}' with lots of platform-dependent defs...
vcvrack_add_plugin(
    SLUG StoneyVCV
    SOURCES "src/plugin.cpp"
    HEADERS "include/plugin.hpp"
    VERSION 2.0.31
)

# This would create an object library; 'plugin' links to it via 'SLUG'...
vcvrack_add_module(
    NAME HP1
    SLUG StoneyVCV
    SOURCES "src/HP1.cpp"
    HEADERS "include/HP1.hpp"
    VERSION 1.0.2
)

# Create as many modules as you wish with it...
vcvrack_add_module(
    NAME LFO
    SLUG StoneyVCV
    SOURCES "src/LFO.cpp"
    HEADERS "include/LFO.hpp"
    VERSION 1.0.1
)

vcvrack_add_module(
    NAME OSC
    SLUG StoneyVCV
    SOURCES "src/OSC.cpp"
    HEADERS "include/OSC.hpp"
    VERSION 1.1.4
)

# Since these functions would create normal CMake targets, the normal CMake functions would still work too..
target_compile_definitions(OSC
    PRIVATE
    "DEBUG"
    "OSC_SPECIFIC_VARIABLE=foo"
)

# Other deps could then be added customarily, on a per-module basis...
find_package(StoneyDSP
    VERSION 1.0.0
    COMPONENTS Core SIMD
)

target_link_libraries(LFO
    PRIVATE
    StoneyDSP::Core
    StoneyDSP::SIMD
)

```

It would be possible to build individual module targets, if desired:

```shell
cmake --build ${PWD}/build --target LFO
```

Of course the excellent thing about functional programming in this way, is that we can offer a function signature (like those above) and do whatever we need to do under the hood regarding implementation changes (SDK version requirements and so forth often change). By updating the function, every plugin/module built this way will receive the updated behaviour automatically. This is a real boon when it comes to different compiler/linker flags etc between different SDK versions, *without* breaking anyone's CMake files, for example...

I have quite a bit of [experience](https://github.com/cmake-js/cmake-js/pull/326) with this [kind of thing](https://github.com/nathanjhood/cmake-js/blob/master/share/cmake/CMakeJS.cmake) - certainly enough to know that it isn't quite as easy as it might sound (input validation is a big one here).

If this post gathers any likes and/or other positive interactions, I will try to make it a higher priority task (I already have made a few rough sketches on local branches...)

*NOTE:* as of writing, all of the above code is pure psuedo-CMake, and likely won't actually work properly in the current form. Feel free to experiment with it though - and let us know of any findings...

**Comment by nathanjhood** [comment]
For the initial commits, I'll be working in `<projectRoot>/share/...` and adding this to my plugin project's `CMAKE_MODULE_PATH`, but this is not the final intention. I'm just working in there while developing this feature, as it saves me having to rebuild the entire Rack SDK CMake target(s) every time I reconfigure my project to test my changes.

Eventually, these changes will be moved into `dep/VCVRack/Rack-SDK/CMakeLists.txt`, and/or possibly elsewhere... (this needs to be detachable so it can be distributed!)

My usual modus operandi; write the pseudo-code first, then write the function signature and make a faux call to it, so you can read the signature semantics (are the parameters in a sensible order? does it scan well? etc), *then* start actually writing the internals...

**Comment by nathanjhood** [comment]
The below code snippet is already compiling succesfully:

```cmake
vcvrack_add_plugin(
    SLUG MySlug
    SOURCES "src/plugin.cpp"
)

vcvrack_add_module(MyModule
    SLUG MySlug
    SOURCES "src/MyModule.cpp"
)
```

It can be found - and tested - in the [demo plugin sub-directory](https://github.com/StoneyDSP/StoneyVCV/blob/feat/vcvrack-cmake-helpers/dep/VCVRack/MyPlugin).

**Comment by nathanjhood** [comment]
Here is the minimum required to compile and build a rack-compatible plugin (a dynamic library) and one linked module (an object library):

```shell
$RACK_DIR/helper.py createplugin MyPlugin
```

```shell
curl https://vcvrack.com/manual/images/MyModule.svg -o res/MyModule.svg
```

```shell
$RACK_DIR/helper.py createmodule MyModule res/MyModule.svg src/MyModule.cpp
```

```.txt
To enable the module, add

	extern Model* modelMyModule;

to plugin.hpp, and add

	p->addModel(modelMyModule);

to the init() function in plugin.cpp.
Source file generated at src/MyModule.cpp
```

```cmake
# CMakeLists.txt
cmake_minimum_required(VERSION 3.14...3.31 FATAL_ERROR)

# Pick up 'VCVRack.cmake' - helper functions
get_filename_component(VCVRACK_DIR "${CMAKE_CURRENT_LIST_DIR}/.." ABSOLUTE)
list(APPEND CMAKE_MODULE_PATH "${VCVRACK_DIR}/Rack-SDK/share/cmake/Modules")

project(MyPlugin)

# Find rack-sdk package from CMAKE_PREFIX_PATH
find_package(rack-sdk 2.5.2 REQUIRED COMPONENTS dep core lib)

# Include VCVRack helpers from CMAKE_MODULE_PATH
include(VCVRack)

vcvrack_add_plugin(
    SLUG MySlug
    SOURCES "src/plugin.cpp" "src/plugin.hpp"
)

vcvrack_add_module(MyModule
    SLUG MySlug
    SOURCES "src/MyModule.cpp"
)
```

```shell
cmake \
    -S ${PWD} \
    -B ${PWD}/build \
    -DCMAKE_TOOLCHAIN_FILE="${VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake"
```

```shell
cmake \
    --build ${PWD}/build \
    --target all
```

**Comment by nathanjhood** [comment]
Here's a more complete example of the functions, which highlights some bugs I'm working on, which don't happen when using the more simplified function call above;

```cmake
# CMakeLists.txt
cmake_minimum_required(VERSION 3.14...3.31 FATAL_ERROR)

project(MyPlugin)

find_package(rack-sdk 2.5.2 REQUIRED COMPONENTS dep core lib)

# the below line won't be required once the design is finished; find_package wll be enough!
# include(VCVRack)

vcvrack_add_plugin(
    SLUG MySlug
    BRAND MyBrand
    HEADERS "include/plugin.hpp"
    SOURCES "src/plugin.cpp"
    VERSION 2.0.1
    SOVERSION 2
)

vcvrack_add_module(MyModule
    SLUG MySlug
    BRAND MyBrand
    SOURCES "src/MyModule.cpp"
    VERSION 0.0.1
    SOVERSION 0
)
```

The above is nice, but unfortunately the module file from the VCV Rack docs' demo module fails to build, since it doesn't carry the `using namespace rack` defined in 'plugin.hpp' - which is real bug-bear for me, using namespaces at global scope in a shared header! eek! Well, we have to assume that a *lot* of Rack plugin devs are following this unfortunate pattern, since probably "it doesn't seem to be a problem". Really, in any case, we want the Rack docs and template code to "just work" unmodified.

I'm trying to find a reliable, and *installable* way to determine in the `module` function whether the `plugin` target carries any interface include directories, and if so, just doing `include_directories` in the module function, to make sure the module target includes the plugin's header file (or it's include path). Brute-forcing it predictably prevents CMake configuring the target successfully (`"this target contains path prefixed in the source directory..."`).

Currently, so long as users either don't use the `HEADERS` param, *or* they use it on *both* targets, *or* they do `include_directories(MyModule ${plugin_include_dirs})`, then they can also do the following:

```cmake
install(TARGETS plugin;MyModule
    EXPORT MyPluginExports
    LIBRARY DESTINATION "${CMAKE_INSTALL_LIBDIR}"
    ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
    RUNTIME DESTINATION "${CMAKE_INSTALL_BINDIR}"
    INCLUDES DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
    FILE_SET plugin_plugin_PUBLIC_HEADERS DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
    FILE_SET plugin_MyModule_PUBLIC_HEADERS DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}"
)

# install export set
install(EXPORT MyPluginExports
    FILE "MyPluginTargets.cmake"
    NAMESPACE MyBrand::MyPlugin::
    DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/MyBrand"
)

# export targets
export(
    TARGETS plugin;MyModule
    FILE "share/cmake/MyPluginTargets.cmake"
    NAMESPACE MyBrand::MyPlugin::
)
```

... which is nice, since it allows us to use CPack to bundle up the distributable in our CI pipelines :D

**Comment by nathanjhood** [comment]
About the JQ stuff... I think it would be nice to *only* check that the mandatory fields in `plugin.json` are defined, and then use a nice little function to compare the target arguments - `VERSION` and `SLUG` and such - against the `plugin.json` and send an error or warning if there is a mismatch.

I *do not* want to get into automating JQ into writing JSON files, nor do I want to use JQ to automatically populate target properties... as a CMake user, that would tying my hands far more than I'd expect it too. I'm free to implement those automations myself in my consumer project, *or* do whatever suites me.

Enforcing such requirements, and hiding standard options from CMake users, creates CMake that people experience difficulties in using.

We can do some more "normal" validations however, to ensure that the plugin's major version matches the Rack SDK, and so forth...

**Comment by nathanjhood** [comment]
> Here's a more complete example of the functions, which highlights some bugs I'm working on, which don't happen when using the more simplified function call above..
>
> The above is nice, but unfortunately...
> 
> I'm trying to find a reliable, and _installable_ way...
> 
> Currently, so long as users either...:
> 

Fixed. The soluton was to check if target `plugin` has a property named `INCLUDE_DIRECTORIES`, and if so, that it doesn't match `""` in a string comparison; in which case, we have include dirs which we just call `target_include_directories(${name} ${_the_iterated_list_of_dirs})`.

I also added an `INCLUDE_DIRS` optional list arg to both sigs, just in case.

New bare minimum:

```cmake
project(MyPlugin)

find_package(rack-sdk COMPONENTS sdk core lib REQUIRED)

vcvrack_add_plugin(
    SLUG MySlug
    HEADERS "src/plugin.hpp" # honestly, headers should be under 'include/'... up to you, though!
    SOURCES "src/plugin.cpp"
)

vcvrack_add_module(MyModule
    SLUG MySlug
    SOURCES "src/MyModule.cpp"
)
```

I'm almost confident enough to start building StoneyVCV with these functions. I am keeping a measure between what I've got in my CMakeLists file vs. what the helpers are doing; the export/install stuff is probably the main thing I've not looked into yet. It won't break to mix the functions with the "normal" CMake stuff anyway :)

**Comment by nathanjhood** [comment]
This is getting beyond a point where tests are needed, and to support that, it might be necessary to start some new repositories for hosting the various pieces:

- `StoneyDSP/StoneyVCV` - My Rack plugin and modules
- `StoneyDSP/StoneyDSP` - My DSP library which my plugin and modules consume
- `StoneyDSP/Rack-SDK-CMake` - The Rack SDK CMakeLists file and some license info 
- `StoneyVCV/vcpkg-registry` - The host of the custom vcpkg port files used for installing the Rack SDK CMake targets with vcpkg
- `StoneyDSP/MyPlugin` - A template-ish project aligned with the VCV Rack Plugin Guide example, being built by the Rack SDK CMake targets (over vcpkg) as a test pipeline

**Comment by nathanjhood** [comment]
The entire "rack sdk helper functions" API is complete; an initial release, at least. It can be polished later.

I have been able to realise the function signatures and methods of operation outlined in this post and elsewhere; I'll not bother repeating myself with more write up here.

Instead, I've added a README where I will begin documenting the new CMake functions.

I haven't bothered linking to those docs here, as we're really close now to being able to part out the entire Rack SDK functionality from this project and putting it into it's own repo, for public consumption.

EDIT: I *have* tested the build on Linux and Windows using VCV Rack 2 Free; it requires running the below commands to manually make the `*.vcvplugin`, taken from `plugin.mk`:

```shell
rm -rf dist
mkdir -p dist/StoneyDSP-StoneyVCV
cp plugin.so dist/StoneyDSP-StoneyVCV/
strip -s dist/StoneyDSP-StoneyVCV/plugin.so
cp -r --parents res LICENSE VERSION presets plugin.json dist/StoneyDSP-StoneyVCV/
cd dist && tar -c StoneyDSP-StoneyVCV | zstd -19 -o "StoneyDSP-StoneyVCV"-"2.0.1"-lin-x64.vcvplugin
mkdir -p "${HOME}/.local/share/Rack2/plugins-lin-x64"
cp dist/*.vcvplugin "${HOME}/.local/share/Rack2/plugins-lin-x64"/
```
